### ISSUE TITLE
Upload covers via copy/paste

### ISSUE THREAD

**Issue by RayBB** [internetarchive/openlibrary#9993]
### Proposal

We would like to be able to upload covers via copy/paste.
Also drag/drop would be nice.
We can no longer support uploading via remote URLs for security purposes.

### Justification

The problem is now people have to download and then upload covers which creates friction.

This would make things easier to upload.

I know our current upload system is in an old iframe and a little hard to develop with (in my opinion)

#### Requirements Checklist

* [ ]

#### Related files

*

#### Stakeholders

* @cdrini 

<hr>

#### Instructions for Contributors

Please [run these commands](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#working-on-your-branch) to ensure your repository is up to date **before** [creating a new branch](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#making-changes-and-creating-a-pull-request) to work on this issue and **each time after** pushing code to Github, because the pre-commit bot may add commits to your PRs upstream.

**Comment by Freso** [comment]
FWIW, MusicBrainz/Cover Art Archive (another IA project) has a long-standing ticket for URL uploading: https://tickets.metabrainz.org/browse/MBS-4641

And there’s a community developed userscript to achieve this functionality: https://github.com/ROpdebee/mb-userscripts/tree/main?tab=readme-ov-file#mb-enhanced-cover-art-uploads (Ray asked me elsewhere to show a demonstration of the userscript in action, but due to the attack(s) on IA, cover art uploading and editing is currently disabled on MB.)

**Comment by cdrini** [comment]
For designs, let's go with something simple like this:

![image](https://github.com/user-attachments/assets/850ea50c-0ede-4c19-ba73-294dd0bbb581)

When an image is pasted, it should then display the image with an "Use this image" button underneath, matching the ui of the IA section.

I think this UX could be simplified/DRYed, but let's do that in a future issue, since I believe getting the JS to work well for the copy/paste use case will be meaty enough to warrant its own issue.

**Comment by Spaarsh** [comment]
@Freso @RayBB @cdrini  @mekarpeles I am willing to work on this issue and have prior experience in vanilla JS. I have went through the [link](https://www.smashingmagazine.com/2018/01/drag-drop-file-uploader-vanilla-js/) provided by earlier.

Additionally, if I may suggest, is it possible for us to add a built-in Object Character Recognition (OCR) to check if the text in the cover image matches the title/author of the book? I found [tesseract.js](https://tesseract.projectnaptha.com/) to be relevant for this use-case. It is an OCR tool that works within the browser (hence client side, no additional load to our servers), in vanilla JS and has support for 100+ languages. This way we can verify if the cover is relevant before it is even uploaded to our servers. At the very least, it can be used to add metadata to the image cover automatically.

**Comment by RayBB** [comment]
@Spaarsh I've assigned you to the issue.
I recommend opening a PR that addresses this issue with image pasting because it's fairly important to get working soon.

Once that is done (or while you're working on it) I'd recommend coming to the weekly call to share your idea for tesseract or write a nice separate feature request ticket. That way you can really explain the benefits and costs and discuss with the community :)

**Comment by RayBB** [comment]
@Spaarsh I'm not the lead but as far as I'm concerned this is certainly on the right track.
The JS and CSS will have to go in their respective files (not just the add.html) but you can learn more about the frontend here
https://github.com/internetarchive/openlibrary/wiki/Frontend-Guide

I'd say, you should be good to open a PR while you wait to hear back from drini about the UI.

**Comment by Spaarsh** [comment]
@RayBB @mekarpeles  @cdrini  @Freso I have added the relevant JavaScript and HTML code to their respective locations according to the front-end guide. The code has been improved upon. 

Here is the demo:
<video src="https://github.com/user-attachments/assets/ac5a7f1b-c20d-47c0-9db8-5e160d558587" width="320" height="240" controls></video>

I will proceed to open a PR now.

Additionally, @RayBB I am yet to flesh out the tesseract.js proposal. However, I have submitted the [Volunteers Form](https://docs.google.com/forms/d/e/1FAIpQLScggx5OI63-oJLES__VSaK83Rlin_m6jfTacjh10wMvHqsZWQ/viewform?usp=pp_url&entry.30454636=%7BEngineering%7D+-+APIs,+data+processing,+bots,+search,+site+performance) as a Software Engineer to join the slack channel. I haven't received any response yet. It was 3 days ago.

**Comment by cdrini** [comment]
Slack invite sent :+1:

**Comment by Spaarsh** [comment]
Received. Thank you!

**Comment by Freso** [comment]
Sorry for off-topic, but please stop @’ing me on every comment. I have no specific stake in this issue. I already get e-mails for all new activity for issues I’m subscribed to. Thank you, and keep up the good work. ❤️

**Comment by RayBB** [comment]
I updated the lead to be @jimchamp because he's assigned to the PR now and I wanna get this one out of limbo. It would be a huge time saver!

**Comment by jimchamp** [comment]
I think that we'll be able to merge the linked PR after one more push.
